libdevice/cmake/modules/SYCLLibdevice.cmake

```
@@ -29,152 +20,138 @@ set(compile_opts
-                           -o ${obj_binary_dir}/libsycl-fallback-complex-fp64.${lib-suffix}
-                   MAIN_DEPENDENCY fallback-complex-fp64.cpp
-                   DEPENDS device_math.h device_complex.h device.h clang clang-offload-bundler
-                   VERBATIM)
-
-add_custom_command(OUTPUT ${spv_binary_dir}/libsycl-fallback-cmath.spv
-                   COMMAND ${clang} -S -fsycl-device-only -fno-sycl-use-bitcode
-                           ${compile_opts}
-                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath.cpp
-                           -o ${spv_binary_dir}/libsycl-fallback-cmath.spv
-                   MAIN_DEPENDENCY fallback-cmath.cpp
-                   DEPENDS device_math.h device.h clang llvm-spirv
-                   VERBATIM)
-
-add_custom_command(OUTPUT ${obj_binary_dir}/libsycl-fallback-cmath.${lib-suffix}
-                   COMMAND ${clang} -fsycl -c
-                           ${compile_opts} ${sycl_targets_opt}
-                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath.cpp
-                           -o ${obj_binary_dir}/libsycl-fallback-cmath.${lib-suffix}
-                   MAIN_DEPENDENCY fallback-cmath.cpp
-                   DEPENDS device_math.h device.h clang clang-offload-bundler
-                   VERBATIM)
-
-add_custom_command(OUTPUT ${spv_binary_dir}/libsycl-fallback-cmath-fp64.spv
-                   COMMAND ${clang} -S -fsycl-device-only -fno-sycl-use-bitcode
-                           ${compile_opts}
-                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath-fp64.cpp
-                           -o ${spv_binary_dir}/libsycl-fallback-cmath-fp64.spv
-                   MAIN_DEPENDENCY fallback-cmath-fp64.cpp
-                   DEPENDS device_math.h device.h clang llvm-spirv
-                   VERBATIM)
-
-add_custom_command(OUTPUT ${obj_binary_dir}/libsycl-fallback-cmath-fp64.${lib-suffix}
-                   COMMAND ${clang} -fsycl -c
-                           ${compile_opts} ${sycl_targets_opt}
-                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath-fp64.cpp
-                           -o ${obj_binary_dir}/libsycl-fallback-cmath-fp64.${lib-suffix}
-                   MAIN_DEPENDENCY fallback-cmath-fp64.cpp
-                   DEPENDS device_math.h device.h clang clang-offload-bundler
-                   VERBATIM)
+function(BuildSYCLDeviceLib)
+  cmake_parse_arguments(DeviceLib "" "Path;Target;Source" "Depends" ${ARGN})
+  add_custom_command(OUTPUT ${DeviceLib_Path}
+                     COMMAND ${clang} -fsycl -fsycl-device-only -emit-llvm
+                             ${compile_opts} -fsycl-targets=${DeviceLib_Target}
+                             ${CMAKE_CURRENT_SOURCE_DIR}/${DeviceLib_Source}
+                             -o ${DeviceLib_Path} -fno-sycl-device-lib=all
+                     MAIN_DEPENDENCY ${DeviceLib_Source}
+                     DEPENDS ${DeviceLib_Depends}
```

I think we can move common dependencies to the function.
```suggestion
                     DEPENDS ${DeviceLib_Depends} device.h clang
```

---

Hi, @bader 
It seems that the  change is not related to "-mlink*" patch, I will create a new PR to clean the CMakeFile to fix all general defects you have pointed out. Is it OK?
Thanks very much.

---

Some of my comments highlight changes also not related to "-mlink*" patch. E.g. we have a few cmake functions to avoid code duplication. This patch removes some of them and duplicate a lot of code instead.

I agree that this is not critical issue and it can be addressed separately, but I would prefer if all comments are addressed before the merge. If you have a tight deadline, I'm okay to address cmake related comments separately.

---

Fixed.